From e2a411384ae17472d3af9fa86013d5c5afab8149 Mon Sep 17 00:00:00 2001
From: Alexander Sosedkin <asosedkin@redhat.com>
Date: Fri, 11 Sep 2026 16:14:22 +0200
Subject: [PATCH] secrets: pass actual zeroes in _tls13_init_secret2

https://datatracker.ietf.org/doc/html/rfc8446#section-7.1
states that key schedule should start with "0", defined as
> a string of Hash.length bytes set to zero.
Passing "" to gnutls_hmac_fast() is computationally equivalent,
but trips a FIPS indicator because this API has a length check.

A third option would've been to call _gnutls_hkdf_extract()
and handle the indicator state like the other functions do,
but this feels a different kind of quirky
and starts hitting a different set of approved algorithm checks,
so I went for a minimal change instead.

Signed-off-by: Alexander Sosedkin <asosedkin@redhat.com>
---
 lib/secrets.c | 6 +++++-
 1 file changed, 5 insertions(+), 1 deletion(-)

diff --git a/lib/secrets.c b/lib/secrets.c
index e6f82c2028..44449f3566 100644
--- a/lib/secrets.c
+++ b/lib/secrets.c
@@ -45,6 +45,7 @@ int _tls13_init_secret2(const mac_entry_st *prf, const uint8_t *psk,
 			size_t psk_size, void *out)
 {
 	char buf[128];
+	char salt[MAX_HASH_SIZE];
 
 	if (unlikely(prf == NULL))
 		return gnutls_assert_val(GNUTLS_E_INTERNAL_ERROR);
@@ -59,7 +60,10 @@ int _tls13_init_secret2(const mac_entry_st *prf, const uint8_t *psk,
 		psk = (uint8_t *)buf;
 	}
 
-	return gnutls_hmac_fast(prf->id, "", 0, psk, psk_size, out);
+	/* passing an empty salt trips FIPS indicator, so pass real zeroes */
+	memset(salt, 0, prf->output_size);
+	return gnutls_hmac_fast(prf->id, salt, prf->output_size, psk, psk_size,
+				out);
 }
 
 /* HKDF-Extract(Prev-Secret, key) */
-- 
2.55.0

